File: fe1-web/src/model/network/method/message/data/chirp/NotifyAddChirp.ts
import { Hash, Timestamp } from 'model/objects';
import { ProtocolError } from 'model/network/ProtocolError';
import { validateDataObject } from 'model/network/validation';
import { ActionType, MessageData, ObjectType } from '../MessageData';
import { checkTimestampStaleness } from '../Checker';

/** Data sent to broadcast a chirp to the general social channel */
export class NotifyAddChirp implements MessageData {
  public readonly object: ObjectType = ObjectType.CHIRP;

  public readonly action: ActionType = ActionType.NOTIFY_ADD;

  // The ID of the chirp
  public readonly chirp_id: Hash;

  // The channel to which the chirp has been posted
  public readonly channel: string;

  // The timestamp at which the chirp has been posted
  public readonly timestamp: Timestamp;

  constructor(msg: Partial<NotifyAddChirp>) {
    if (!msg.chirp_id) {
      throw new ProtocolError('Undefined \'id\' parameter encountered during \'NotifyAddChirp\'');
    }
    this.chirp_id = msg.chirp_id;

    if (!msg.timestamp) {
      throw new ProtocolError('Undefined \'timestamp\' parameter encountered during \'NotifyAddChirp\'');
    }
    checkTimestampStaleness(msg.timestamp);
    this.timestamp = msg.timestamp;

    if (!msg.channel) {
      throw new ProtocolError('Undefined \'channel\' parameter encountered during \'NotifyAddChirp\'');
    }
    this.channel = msg.channel;
  }

  /**
   * Creates an NotifyAddChirp object from a given JSON object.
   *
   * @param obj - The given JSON object
   */
  public static fromJson(obj: any): NotifyAddChirp {
    const { errors } = validateDataObject(ObjectType.CHIRP, ActionType.NOTIFY_ADD, obj);

    if (errors !== null) {
      throw new ProtocolError(`Invalid receive chirp\n\n${errors}`);
    }

    return new NotifyAddChirp({
      ...obj,
      chirp_id: new Hash(obj.chirp_id),
      timestamp: new Timestamp(obj.timestamp),
    });
  }
}

File: be2-scala/src/main/scala/ch/epfl/pop/model/network/method/message/data/socialMedia/NotifyAddChirp.scala
package ch.epfl.pop.model.network.method.message.data.socialMedia

import ch.epfl.pop.json.MessageDataProtocol._
import ch.epfl.pop.model.network.Parsable
import ch.epfl.pop.model.network.method.message.data.{ActionType, MessageData, ObjectType}
import ch.epfl.pop.model.objects.{Channel, Hash, Timestamp}
import spray.json._

final case class NotifyAddChirp(
    chirp_id: Hash,
    channel: Channel,
    timestamp: Timestamp
) extends MessageData {
  override val _object: ObjectType = ObjectType.chirp
  override val action: ActionType = ActionType.notify_add
}

object NotifyAddChirp extends Parsable {
  def apply(
      chirp_id: Hash,
      channel: Channel,
      timestamp: Timestamp
  ): NotifyAddChirp = new NotifyAddChirp(chirp_id, channel, timestamp)

  override def buildFromJson(payload: String): NotifyAddChirp = payload.parseJson.asJsObject.convertTo[NotifyAddChirp]
}
